### ISSUE TITLE
Future of ted-components, ted-bootstrap

### ISSUE THREAD

**Issue by samselikoff** [tedconf/code-style-guides#26]
Hey Emberinos. How do we want to develop shared components?

Currently we have [ember-cli-ted-bootstrap](https://github.com/tedconf/ember-cli-ted-bootstrap) and [ember-cli-ted-components](https://github.com/tedconf/ember-cli-ted-components/).

First, both have a `<ted-panel>` that should not be used since we now have `<collapsible-panel>`. But more generally, how should we be working on these going forward? Should we break them out into individual addons? For example, I want to improve (and rename!) `<ted-flaps>`. I want to add some animation, and I want us to figure out what the behavior should be on small screens. Should we pull this out into a separate addon?

Thoughts?

**Comment by brenna** [comment]
Instinct is to split out as much as we can:
1) easier to maintain and version without in-app conflicts
2) more potential for open source sharing

There might be some room to keep a small set of core baseline/getting-started styles in a bootstrap like repo.

**Comment by samselikoff** [comment]
I agree. @ryanto what do you think? If we all agree I'll kick it off by creating a new `<ted-flaps>` (after we decide what to call it).

**Comment by brenna** [comment]
Yeah, anything but "flaps" :)

**Comment by ryanto** [comment]
Split out + open source.

**Comment by samselikoff** [comment]
Sweet!

While I have you all here - what should `<ted-flaps>` be called?

We have `<ted-navbar>`. Flaps is really a subnav.

 - `<ted-subnav>`
 - `<ted-subnavbar>`
 - `<ted-nav>` (it's just our standard nav, after all, and it's not a bar. kinda makes sense)
 - others?

**Comment by brenna** [comment]
+1 <ted-nav>

**Comment by redoPop** [comment]
Re. naming the navigation component: `ted-nav` sounds good to me too! Some historical context: "ted-flaps" borrowed its name from ted_bootstrap's "nav-flaps", which was chosen as a class name because Bootstrap is already using [`nav`](http://getbootstrap.com/components/#nav), [`navbar`](http://getbootstrap.com/components/#navbar), and other more reasonable naming options. Namespacing the componentized version with the `ted-` prefix (e.g., `ted-nav`) is a better solution for avoiding naming conflicts. Only downside is it's not very flappy. 🐦

---

Re. the broader topic being discussed here: I think the relationship between Ember component libraries and Bootstrap (or [ted_bootstrap](https://github.com/tedconf/ted_bootstrap)) is something of an elephant in this room.

Originally, ted_bootstrap was intended to be drop-in CSS for any of our internal tools, regardless of JS framework. To that end, it provides standard header and footer markup and styles, including the "nav-flaps". As I see it, `ted-nav` introduces an Ember dependency to that standard header, formalizing Ember as a dependency wherever it's used.

A reciprocal dependency _upon_ Bootstrap exists within the Ember components: ember-cli-ted-bootstrap's components are obviously intended for use with Bootstrap, but ember-cli-ted-components also assumes or mentions Bootstrap in several places, and some components are built on styles extracted from ted_bootstrap.

Assuming this is true most of the time, there are a couple of things I'm wondering about here:

* When breaking these components out into their own public repos, we'll probably be tempted to duplicate, re-engineer, or otherwise avoid relying on Bootstrap in an effort to make the outcome more broadly applicable to other developers. How should we balance that desire – which I think is legitimate and justifiable – with the fact that we typically work in the context of Bootstrap?
* If Ember is to become a dependency wherever Bootstrap is used, should we start removing styles from ted_bootstrap as they're supplanted by Ember components? e.g. should the addition of the `ted-nav` component be accompanied by a removal of `nav-flaps` from ted_bootstrap?

**Comment by samselikoff** [comment]
Thanks for your input Joe. Sorry if it seemed I was leaving you out!

This was actually intended to be a much narrower discussion, just about how we were breaking up *ember* components within our two existing addons, `ember-cli-ted-bootstrap` and `ember-cli-ted-components`, and whether we should leave them grouped together or separate them one by one. This isn't meant to discuss any changes that would affect non-Ember apps.

**Comment by samselikoff** [comment]
To clarify, `<ted-nav>` would be an ember component in an `ember-ted-nav` library. It will assume the host app is using ted-bootstrap via ember-cli-ted-bootstrap.

**Comment by redoPop** [comment]
Oh, no, not at all! I understood your intent, but the thing I'm pointing out is a budding reciprocal dependency between Bootstrap and our Ember components.

TBH I think there are very few (if any!) consumers of ted_bootstrap that aren't already Ember projects, and although it's currently a separate project `ember-ted-nav` and its component `<ted-nav>` will likely extend and then supplant the styles behind ted_bootstrap's `nav-flaps` in a similarly de-facto manner.

**Comment by samselikoff** [comment]
Ryan's actually working on a rails app now that uses ted_bootstrap, so I think it's okay that we leave them separate.

For now, let's make `ember-ted-nav` (and its `<ted-nav>` component) a new addon. It will wrap the current flaps markup from ted_bootstrap, so it will still depend on ted_bootstrap, which I think is fine. It's just a convenience for us TED developers. Eventually, as we start to use it, and if we like it, we can push changes upstream to ted_bootstrap to rename flaps to nav. Do you agree?

Regarding your first bullet point above, [searchable-select](tedconf.github.io/ember-searchable-select/) is its own generic addon/component that has nothing to do with Bootstrap, and [collapsible-panel](https://tedconf.github.io/ember-collapsible-panel), which originally wrapped Bootstrap's collapse plugin, is now its own Bootstrap-independent component. So, I would suggest that we move towards making components independent from Bootstrap's styles and/or JS as the need arises. That's essentially what we've done so far, and I think it's fine that other addons (like [ember-ted-gear](http://tedconf.github.io/ember-ted-gear/)) are still wrappers for Bootstrap's JS.

**Comment by redoPop** [comment]
Ooh. That ted_bootstrap is under active use in the wild definitely changes things. (Damn it, Ryan!)

Maybe I'm just imagining problems here, but there's probably a conflict between these two goals:

> I want to add some animation, and I want us to figure out what the behavior should be on small screens.
> …
> It will wrap the current flaps markup from ted_bootstrap, so it will still depend on ted_bootstrap, which I think is fine.

Creating Ember components that depend on / correspond to ted_bootstrap is fine, but in this case particularly I suspect the Ember version's enhancements will quickly outgrow Bootstrap's styles and it will likely end up as more of a reimplementation than a wrapper. That's not bad per se, so long as it's understood and expected. (e.g. we may need to have cross-referenced documentation in the Ember component and the ted_bootstrap docs pointing to the alternative version of the widget.)

> Eventually, as we start to use it, and if we like it, we can push changes upstream to ted_bootstrap to rename flaps to nav.

There's a naming conflict with "nav" in Bootstrap ([per my notes following "some historical context" in this comment](https://github.com/tedconf/code-style-guides/issues/26#issuecomment-169161244)). We could rename the Bootstrap class to "ted-nav" but that might conflict the Ember component's class name.

Other than that detail, this approach works for me!

**Comment by samselikoff** [comment]
Ahh, okay I see what you're saying now. The animation bit is entirely additional to the CSS/HTML, because we were toying with doing something like [this](http://hoover-office.ted.com/?id=27) (you need to be on vpn to see that) for the Ember component. This type of animation would be ember-only, since it's stateful. So it would never make it back into ted_bootstrap. (Of course, unless we changed it to stateless/css-only animation - but that'd be a stylistic choice, and it's still all early/experimental stuff).

The small-screen behavior could definitely make it into ted_bootstrap, though, if it's html/css based (which it probably will/should be). How about we have that discussion in ted_bootstrap, but split the ember component out so we can change the name/api a bit, and play with stateful animation?

Thanks for taking the time to bring up these points.

**Comment by redoPop** [comment]
Awesome; this sounds great!
